azure/Guidelines.md

```diff
@@ -0,0 +1,861 @@
+
+#### Direct Endpoint URLs
+
+> :heavy_check_mark: **YOU MAY** support a direct endpoint URL for performance/routing:
+```text
+https://<tenant>-<service-root>.<service>.<cloud>/...
+```
+
+Examples:
+  - Request URL: `https://blobstore.azure.net/contoso.com/account1/container1/blob2`
+  - Response header ([RFC2557](https://datatracker.ietf.org/doc/html/rfc2557#section-4)): `content-location : https://contoso-dot-com-account1.blobstore.azure.net/container1/blob2`
+  - GUID format: `https://00000000-0000-0000-C000-000000000046-account1.blobstore.azure.net/container1/blob2`
+
+> :white_check_mark: **DO** return URLs in response headers/bodies in a consistent form regardless of the URL used to reach the resource. Either always a UUID for `<tenant>` or always a single verified domain.
+
+> :heavy_check_mark: **YOU MAY** use URLs as values
+```
+https://api.contoso.com/items?url=https://resources.contoso.com/shoes/fancy
+```
+
+### HTTP Request / Response Pattern
+The HTTP Request / Response pattern dictates how your API behaves. For example: POST methods that create resources must be idempotent, GET method results may be cached, the If-Modified and etag headers offer optimistic concurrency. The URL of a service, along with its request/response bodies, establishes the overall contract that developers have with your service. As a service provider, how you manage the overall request / response pattern should be one of the first implementation decisions you make.
+
+Cloud applications embrace failure. Therefore, to enable customers to write fault-tolerant applications, <b>all</b> service operations (including POST) <b>must</b> be idempotent. Implementing services in an idempotent manner, with an "exactly once" semantic, enables developers to retry requests without the risk of unintended consequences.
+
+#### Exactly Once Behavior = Client Retries & Service Idempotency
+> :white_check_mark: **DO** ensure that __all__ HTTP methods are idempotent.
+ 
+> :ballot_box_with_check: **YOU SHOULD** use PUT or PATCH to create a resource as these HTTP methods are easy to implement, allow the customer to name their own resource, and are idempotent.
+ 
+> :heavy_check_mark: **YOU MAY** use POST to create a resource but you must make it idempotent and, of course, the response __must__ return the URL of the created resource with a 201-Created. One way to make POST idempotent is to use the Repeatability-Request-ID & Repeatability-First-Sent headers (See [Repeatability of requests](#Repeatability-of-requests)).
+
+##### HTTP Return Codes
+> :white_check_mark: **DO** adhere to the return codes in the following table when the method is successful:
+
+Method | Description | Response Status Code 
+-------|-------------|---------------------
+PATCH  | Create/Modify the resource with JSON Merge Patch | 200-OK, 201-Created
+PUT    | Create/Replace the *whole* resource | 200-OK, 201-Created 
```

Comment:
No 202 LRO?

Reply:
Yes on supporting 202 on PUT/POST/DELETE.

In the LRO section, we will add additional guidance on these.